tensorflow_addons/losses/multi_similarity_loss.py

```diff
@@ -0,0 +1,169 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements Multi Similarity loss."""
+
+import tensorflow as tf
+from tensorflow_addons.utils.types import FloatTensorLike, TensorLike
+from typeguard import typechecked
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiSimilarityLoss(tf.keras.losses.Loss):
+    """Implements the multi similarity loss function.
+
+    Multi Similarity Loss was first introduced in "Multi-Similarity Loss with
+    General Pair Weighting for Deep Metric Learning" (https://arxiv.org/pdf/1904.06627.pdf).
+    This loss is implemented in two iterative steps (i.e., mining and weighting).
+    This allows it to fully consider three similarities for pair weighting,
+    providing a more principled approach for collecting and weighting informative pairs.
+    Finally, the proposed MS loss obtains new state-of-the-art performance on four image
+    retrieval benchmarks, where it outperforms the most recent approaches.
+
+    Usage:
+
+    ```python
+    msl = tfa.losses.MultiSimilarityLoss()
+    loss = msl(
+      [[0.97], [0.91], [0.03]],
+      [[1.0], [1.0], [0.0]])
+    print('Loss: ', loss.numpy())  # Loss:  0.009241962
+    ```
+    Usage with tf.keras API:
+
+    ```python
+    model = tf.keras.Model(inputs, outputs)
+    model.compile('sgd', loss=tf.keras.losses.MultiSimilarityLoss())
+    ```
+
+    Args:
+      alpha: Hyper-parameter similar as Binomial Deviance Loss.
+      beta: Hyper-parameter similar as Binomial Deviance Loss.
+      lamb: Hyper-parameter similar as Binomial Deviance Loss.
+      eps: very small positive number to prevent devide by zero error.
+      ms_mining: bool.
+
+    Returns:
+      Loss float `Tensor`.
+    """
+
+    @typechecked
+    def __init__(
+        self,
+        alpha: FloatTensorLike = 2.0,
+        beta: FloatTensorLike = 50.0,
+        lamb: FloatTensorLike = 1.0,
+        eps: FloatTensorLike = 0.1,
+        ms_mining: bool = False,
+        from_logits: bool = False,
+        name: str = "multi_similarity_loss",
+    ):
+        super().__init__(name=name)
+        self.alpha = alpha
+        self.beta = beta
+        self.lamb = lamb
+        self.eps = eps
+        self.ms_mining = ms_mining
+        self.from_logits = from_logits
+
+    def call(self, y_true, y_pred):
+        return multi_similarity_loss(
+            y_true,
+            y_pred,
+            self.alpha,
+            self.beta,
+            self.lamb,
+            self.eps,
+            self.ms_mining,
+            self.from_logits,
+        )
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+@tf.function
+def multi_similarity_loss(
+    y_true: TensorLike,
+    y_pred: TensorLike,
+    alpha: FloatTensorLike = 2.0,
+    beta: FloatTensorLike = 2.0,
+    lamb: FloatTensorLike = 1.0,
+    eps: FloatTensorLike = 0.1,
```

Comment:
our public API all use `epsilon` (core TF as well), please change it to `epsilon` for consistency. Also, `0.1` is not a very **small positive number**. Is there any reason we set it to such large value?